### PR TITLE
Added High Order and Low Order RK methods

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [X] Any new documentation only uses public API
  
## Additional context

solves a part of #2177

### CODE REVIEW

File: lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+alg_order(alg::SplitEuler) = 1
+alg_order(alg::Heun) = 2
+alg_order(alg::Ralston) = 2
+alg_order(alg::Midpoint) = 2
+alg_order(alg::BS3) = 3
+alg_order(alg::OwrenZen3) = 3
+alg_order(alg::BS5) = 5
+alg_order(alg::OwrenZen4) = 4
+alg_order(alg::OwrenZen5) = 5
+alg_order(alg::Tsit5) = 5
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl:11]: We will want to make Tsit5 be its own set.

File: lib/OrdinaryDiffEqLowOrderRK/test/runtests.jl [hunk 2]
```diff
@@ -0,0 +1,3 @@
+using SafeTestsets
+
+@time @safetestset "Extrapolation Tests" include("owrenzen_tests.jl")
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowOrderRK/test/runtests.jl:3]:
```suggestion
@time @safetestset "OwrenZen Tests" include("owrenzen_tests.jl")
```

File: src/algorithms.jl [hunk 3]
```diff
@@ -818,8 +593,7 @@ for Alg in [:Exp4, :EPIRK4s3A, :EPIRK4s3B, :EPIRK5s3, :EXPRB53s3, :EPIRK5P1, :EP
             iop)
     end
 end
-struct SplitEuler <:
```
Comment (ChrisRackauckas) [inline src/algorithms.jl:821]: why was this removed?

File: test/runtests.jl [hunk 4]
```diff
@@ -224,6 +223,10 @@ end
         @time @safetestset "Linear-Nonlinear Krylov Methods Tests" include("algconvergence/linear_nonlinear_krylov_tests.jl")
     end
 
+    if !is_APPVEYOR && GROUP == "OwrenZen"
```
Comment (ChrisRackauckas) [inline test/runtests.jl:226]:
```suggestion
    if !is_APPVEYOR && GROUP == "LowOrderRK"
```

File: lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl (hunk 1 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl:11]: Ok so for this I will make a separate OrdinaryDiffEqTsit inside lib

File: src/algorithms.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/algorithms.jl:821]: It should end up with the split methods group, with CNAB2 and such. What's the dependency?

File: lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply lib/OrdinaryDiffEqLowOrderRK/src/alg_utils.jl:11]: OrdinaryDiffEqTsit5. Since it's the most common method and it's commonly used alone, making it as lean as possible is a virtue. Thanks.

**Comment by ParamThakkar123** [comment]
> Can we take this moment to split `@time @safetestset "Convergence Tests" include("algconvergence/ode_convergence_tests.jl")` into its relevant packages? A lot of methods just dumped their convergence test there, so it mixes a few, but most are these low and high order RKs.

We can start working on splitting this one then

File: src/algorithms.jl (hunk 3 above)
Comment (ParamThakkar123) [reply src/algorithms.jl:821]: image

when I went through low_order_rk_caches.jl file I saw the splitEuler Cache here so I thought that splitEuler might be an algorithm of low_order_rk. That was the reason

**Comment by ParamThakkar123** [comment]
image

@ChrisRackauckas Why did this happen ?? We didn't get these warnings previously

File: src/algorithms.jl (hunk 3 above)
Comment (ChrisRackauckas) [reply src/algorithms.jl:821]: That was a good guess but ultimately not the right one. There should be a SplitMethods which has CNAB2 and such, and SplitEuler is kind of a smaller test method that goes with that set.

**Comment by ChrisRackauckas** [comment]
Fixed that.

File: lib/OrdinaryDiffEqTsit5/src/OrdinaryDiffEqTsit5.jl [hunk 5]
```diff
@@ -0,0 +1,24 @@
+module OrdinaryDiffEqTsit5
+
+import OrdinaryDiffEq: alg_order, alg_stability_size, explicit_rk_docstring,
+                       OrdinaryDiffEqAdaptiveAlgorithm, OrdinaryDiffEqMutableCache, alg_cache,
+                       OrdinaryDiffEqConstantCache, @fold,
+                       constvalue, @unpack, perform_step!, calculate_residuals, @cache,
+                       calculate_residuals!, _ode_interpolant, _ode_interpolant!
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqTsit5/src/OrdinaryDiffEqTsit5.jl:7]:
```suggestion
                       calculate_residuals!, _ode_interpolant, _ode_interpolant!,
                       CompiledFloats
```

File: lib/OrdinaryDiffEqLowOrderRK/src/low_order_rk_caches.jl [hunk 6]
```diff
@@ -1535,3 +1485,27 @@ function alg_cache(alg::Alshina6, u, rate_prototype, ::Type{uEltypeNoUnits},
     Alshina6Cache(u, uprev, k1, k2, k3, k4, k5, k6, k7, tmp, tab, alg.stage_limiter!,
         alg.step_limiter!, alg.thread)
 end
+
+@cache struct FunctionMapCache{uType, rateType} <: OrdinaryDiffEqMutableCache
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowOrderRK/src/low_order_rk_caches.jl:1489]: FunctionMap is separate from the LowOrderRK methods because it acts on discrete problems instead of ODEs.

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqLowOrderRK/src/low_order_rk_caches.jl:1489]: Should I make it a separate library for this one ?? similar to Tsit ??

Comment (ChrisRackauckas) [reply lib/OrdinaryDiffEqLowOrderRK/src/low_order_rk_caches.jl:1489]: Yes

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas  `alg_cache`, `initialize!` and other methods are showing method errors from various solvers saying No match for the method. Any fixes for this one that I can try ??

**Comment by ChrisRackauckas** [comment]
I put a commit in that should handle the functionmap dispatch issues with its moving. I'll see what the next ones are.

File: lib/OrdinaryDiffEqLowOrderRK/src/OrdinaryDiffEqLowOrderRK.jl [hunk 7]
```diff
@@ -0,0 +1,33 @@
+module OrdinaryDiffEqLowOrderRK
+
+import OrdinaryDiffEq: alg_order, isfsal, beta2_default, beta1_default, alg_stability_size,
+                       ssp_coefficient, OrdinaryDiffEqAlgorithm, OrdinaryDiffEqExponentialAlgorithm,
+                       explicit_rk_docstring, trivial_limiter!, OrdinaryDiffEqAdaptiveAlgorithm,
+                       unwrap_alg, @unpack, initialize!, perform_step!, calculate_residuals,
+                       calculate_residuals!, _ode_addsteps!, @OnDemandTableauExtract, constvalue,
+                       OrdinaryDiffEqMutableCache, uses_uprev, OrdinaryDiffEqConstantCache, @fold,
+                       @cache, CompiledFloats
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowOrderRK/src/OrdinaryDiffEqLowOrderRK.jl:9]:
```suggestion
                       @cache, CompiledFloats, alg_cache
```

File: src/OrdinaryDiffEq.jl [hunk 8]
```diff
@@ -170,32 +160,15 @@ include("cache_utils.jl")
...
-include("dense/stiff_addsteps.jl")
-include("dense/low_order_rk_addsteps.jl")
-include("dense/high_order_rk_addsteps.jl")
 
 include("derivative_utils.jl")
-include("nordsieck_utils.jl")
-include("adams_utils.jl")
 include("derivative_wrappers.jl")
 include("iterator_interface.jl")
 include("constants.jl")
```
Comment (ChrisRackauckas) [inline src/OrdinaryDiffEq.jl:171]:
```suggestion
include("constants.jl")
include("solve.jl")
```

File: lib/OrdinaryDiffEqRosenbrock/src/OrdinaryDiffEqRosenbrock.jl [hunk 9]
```diff
@@ -0,0 +1,29 @@
+module OrdinaryDiffEqRosenbrock
+
+import OrdinaryDiffEq: alg_order, alg_adaptive_order, isWmethod, isfsal, _unwrap_val, rosenbrock_wanner_docstring,
+                       DEFAULT_PRECS, OrdinaryDiffEqRosenbrockAlgorithm, @cache, alg_cache, initialize!, @unpack,
+                       calc_W, calculate_residuals!, calc_rosenbrock_differentiation!, OrdinaryDiffEqMutableCache,
+                       build_J_W, UJacobianWrapper, OrdinaryDiffEqConstantCache, _ode_interpolant, _ode_interpolant!,
+                       _vec, _reshape, perform_step!
+using TruncatedStacktraces, MuladdMacro, FastBroadcast, DiffEqBase, RecursiveArrayTools, step_limiter!, dolinsolve
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqRosenbrock/src/OrdinaryDiffEqRosenbrock.jl:8]:
```suggestion
                       _vec, _reshape, perform_step!, trivial_limiter!, dolinsolve
using TruncatedStacktraces, MuladdMacro, FastBroadcast, DiffEqBase, RecursiveArrayTools, 
```

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas in which file is the code for exponential RK written (algorithms and caches) ??

**Comment by ChrisRackauckas** [comment]
linear caches, linear perform step

**Comment by ChrisRackauckas** [comment]
let's try to get this merged before adding some more though. Taking it in steps is always simpler.

**Comment by ChrisRackauckas** [comment]
ExpeontialRK is with linear_nonlinear caches, which is separate from the non-autonomous linear solvers in the linear caches. But let's do this one as a separate PR. There's some other things to do there with dependencies.

**Comment by ChrisRackauckas** [comment]
Note the composite alg doesn't move since it's more of an internal tool.

File: src/solve.jl [hunk 10]
```diff
@@ -27,11 +27,11 @@ function DiffEqBase.__init(
         save_end = nothing,
         callback = nothing,
         dense = save_everystep &&
-                    !(alg isa Union{DAEAlgorithm, FunctionMap}) &&
+                    !(alg isa DAEAlgorithm) && !(prob isa DiscreteProblem) &&
                     isempty(saveat),
         calck = (callback !== nothing && callback !== CallbackSet()) ||
                     (dense) || !isempty(saveat), # and no dense output
-        dt = alg isa FunctionMap && isempty(tstops) ?
+        dt = !dt_required(alg) && isempty(tstops) ?
```
Comment (ChrisRackauckas) [inline src/solve.jl:34]:
```suggestion
        dt = isdiscretealg(alg) && isempty(tstops) ?
```

File: src/solve.jl [hunk 11]
```diff
@@ -130,8 +130,7 @@ function DiffEqBase.__init(
     if (((!(alg isa OrdinaryDiffEqAdaptiveAlgorithm) &&
           !(alg isa OrdinaryDiffEqCompositeAlgorithm) &&
           !(alg isa DAEAlgorithm)) || !adaptive || !isadaptive(alg)) &&
-        dt == tType(0) && isempty(tstops)) &&
-       !(alg isa Union{FunctionMap, LinearExponential})
+        dt == tType(0) && isempty(tstops)) && !dt_required(alg)
```
Comment (ChrisRackauckas) [inline src/solve.jl:133]:
```suggestion
        dt == tType(0) && isempty(tstops)) && dt_required(alg)
```

File: lib/OrdinaryDiffEqNordsieck/src/OrdinaryDiffEqNordsieck.jl [hunk 12]
```diff
@@ -0,0 +1,18 @@
+module OrdinaryDiffEqNordsieck
+
+import OrdinaryDiffEq: alg_order, alg_adaptive_order, qsteady_max_default, get_current_alg_order,
+                       AbstractController, OrdinaryDiffEqAdaptiveAlgorithm, OrdinaryDiffEqAdamsVarOrderVarStepAlgorithm,
+                       alg_cache, OrdinaryDiffEqMutableCache, OrdinaryDiffEqConstantCache, initialize!, @unpack,
+                       initialize!, perform_step!, stepsize_controller!, step_accept_controller!, step_reject_controller!,
+                       calculate_residuals, calculate_residuals!, Tsit5ConstantCache
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqNordsieck/src/OrdinaryDiffEqNordsieck.jl:7]:
```suggestion
                       calculate_residuals, calculate_residuals!, Tsit5ConstantCache, get_current_adaptive_order, ode_interpolant, ode_interpolant!
```

File: src/OrdinaryDiffEq.jl [hunk 13]
```diff
@@ -293,6 +278,36 @@ include("../lib/OrdinaryDiffEqPRK/src/OrdinaryDiffEqPRK.jl")
...
+export AB3, AB4, AB5, ABM32, ABM43, ABM54, VCAB3,
+       VCAB4, VCAB5, VCABM3, VCABM4, VCABM5, VCABM
+
+include("../lib/OrdinaryDiffEqNordsieck/src/OrdinaryDiffEqNordsieck.jl")
+using ..OrdinaryDiffEqNordsieck
+export AN5, JVODE, JVODE_Adams, JVODE_BDF
+
+include("../lib/OrdinaryDiffEqExplicitRK/src/OrdinaryDiffEqExplicitRK.jl")
+using ..OrdinaryDiffEqExplicitRK
+export ExplicitRK
```
Comment (ChrisRackauckas) [inline src/OrdinaryDiffEq.jl:309]:
```suggestion
using ..OrdinaryDiffEqExplicitRK
using ..OrdinaryDiffEqExplicitRK: constructDormandPrince
export ExplicitRK
```

File: lib/OrdinaryDiffEqLowOrderRK/src/OrdinaryDiffEqLowOrderRK.jl [hunk 14]
```diff
@@ -9,6 +9,7 @@ import OrdinaryDiffEq: alg_order, isfsal, beta2_default, beta1_default, alg_stab
                        @cache, CompiledFloats, alg_cache, CompositeAlgorithm, alg_autodiff, copyat_or_push!,
                        AutoAlgSwitch, _ode_interpolant, _ode_interpolant!, full_cache,
                        accept_step_controller, DerivativeOrderNotPossibleError
+                       AutoAlgSwitch, _ode_interpolant, _ode_interpolant!, full_cache, DP8ConstantCache, DP8Cache
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowOrderRK/src/OrdinaryDiffEqLowOrderRK.jl]: No this should not be here.

Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqLowOrderRK/src/OrdinaryDiffEqLowOrderRK.jl]: Removed it

**Comment by ChrisRackauckas** [comment]
Almost there!

**Comment by ChrisRackauckas** [comment]
@oscardssmith I reverted your Rosenbrock commit because it broke some tests.

**Comment by ParamThakkar123** [comment]
Which solvers should I go with now ??

**Comment by ChrisRackauckas** [comment]
IMEXMultistep, Linear, and ExponenitalRK are the last 3 sets.
